### DISCUSSION TITLE
Disassembly View not showing debug details

### DISCUSSION THREAD

**Discussion by xshuqun** [eclipse-cdt/cdt#773] [Q&A] [answered]
Hi All,
I have created my own project, which I had managed to build successfully. I had also added downloading and debugging configuration for my project. The downloading is working well, I was also able to enter debug, but the Disassembly view is not showing any address.
![image](https://github.com/eclipse-cdt/cdt/assets/77948861/f27f4e49-dd41-44dc-bb76-8e7b37efdba6)

I had ensured that I have marked the "Load Symbols".
![image](https://github.com/eclipse-cdt/cdt/assets/77948861/d31ac257-a401-4dd7-8d2e-c51db459c7f1)

Does anyone know what could be the issue here?

Thank you

**Comment by jonahgraham** [comment]
The disassembly view in your screenshot shows "no debug context" which means that in the debug view that nothing is selected, or if selected it isn't suspended. Here is a screencast of me selecting and unselecting the frame in the debug view.

![Peek 2024-04-30 08-47](https://github.com/eclipse-cdt/cdt/assets/679236/ddf337fd-5a02-48ca-a846-cc1abd3773a0)

**Reply by xshuqun** [reply]
Thanks @jonahgraham , I do see some addresses now in the Disassembly view, however, it doesn't seem to be correct. I tried triggering breakpoint at different area, and selecting different frames, but it is still showing the same view as below.
![image](https://github.com/eclipse-cdt/cdt/assets/77948861/d3f76026-42b5-47ab-8427-b6cc0ef8ea10)
Thank you.

**Reply by xshuqun** [reply]
Hi @jonahgraham , could i get some help with this?
Any idea how should i check on this?

**Reply by jonahgraham** [reply]
The disassembly view is showing the "correct" thing as far as it understands it. GDB/JLink are reporting back the current address as 0x0 (from _main() as main.c:92 **0x0**_). So the issue here looks like something else. Not sure how I can help on this one. Is the problem you are having only with disassembly view? Is everything else working ok?

**Reply by xshuqun** [reply]
Yes, the only problem is with the disassembly view, the other functions are working perfectly well. But I think the dissembly view is essential for debugging, but I am uncertain why the addresses are incorrectly displayed.

**Reply by jonahgraham** [reply]
Sorry but some clarification needed. You said " other functions are working perfectly well" - but is the "main() as main.c:92 0x0" in the *Debug* view correct?

If it is correct, then the disassembly view is displaying the correct (as we understand it) addresses.

If the "main() as main.c:92 0x0" is showing the incorrect address value, it implies that gdb is reporting back that main.c:92 is at address 0x0. That is what we need to resolve. 

Does the 0x0 change when you single step your code? I would expect it to increment.

Can you check the [gdb traces (instructions on this link)](https://github.com/eclipse-cdt/cdt/tree/main/FAQ#ive-been-asked-for-gdb-traces-where-can-i-find-them) to see if gdb is indeed reporting back 0x0 or if CDT is misinterpretting something as 0x0. Here is what I see, as you can see the red circled address in the traces equals what is shown in the Debug view and disassembly view:

![image](https://github.com/eclipse-cdt/cdt/assets/679236/643bb657-7375-4c76-88aa-3bb36a65a22b)

**Reply by xshuqun** [reply]
The "main() as main.c:92 0x0" doesn't seem to be correct, as i resume the programs and stops again, it remains at 0x0, which is weird.
This is the console output when i try to halt the problem.
For your information, I am using Jlink Segger as my debugger.

> 706,395 *stopped,reason="signal-received",signal-name="SIGTRAP",signal-meaning="Trace/breakpoint trap",frame={func="xTaskResumeAll",args=[],file="SDK_PATH",fullname="SDK_PATH",line="1641"},thread-id="1",stopped-threads="all"
706,395 (gdb) 
706,395 118^done,new-thread-id="1",frame={level="0",func="xTaskResumeAll",args=[],file="SDK_PATH",fullname="SDK_PATH",line="1641"}
706,395 (gdb) 
706,399 119-stack-info-depth --thread 1 11
706,505 119^done,depth="4"
706,505 (gdb) 
706,509 120-stack-select-frame --thread 1 0
706,509 120^done
706,509 (gdb) 
706,573 121-list-thread-groups i1
706,573 122-list-thread-groups
706,574 121^done,threads=[{id="1",target-id="Thread 57005",frame={level="0",func="xTaskResumeAll",args=[],file="SDK_PATH",fullname="SDK_PATH",line="1641"},state="stopped"}]
706,574 (gdb) 
706,574 122^done,groups=[{id="i1",type="process",pid="42000",executable="SDK_PATH/Debug/application.axf"}]
706,574 (gdb) 
706,603 123-stack-list-locals --thread 1 --frame 0 1
706,603 123^done,locals=[{name="pxTCB",value="<optimized out>"},{name="xAlreadyYielded",value="0"}]
706,603 (gdb) 
706,611 124-stack-list-frames --thread 1
706,612 124^done,stack=[frame={level="0",func="xTaskResumeAll",file="SDK_PATH",fullname="SDK_PATH",line="1641"},frame={level="1",func="prvIdleTask",file="SDK_PATH",fullname=[message truncated to 5 lines. More lines can be set in the Preference->Debug->GDB eclipse preference page.]
706,612 (gdb) 
706,614 125-var-create --thread 1 --frame 0 - * pxTCB
706,614 126-var-create --thread 1 --frame 0 - * xAlreadyYielded
706,614 125^done,name="var1",numchild="8",value="<optimized out>",type="TCB_t *",thread-id="1",has_m\
ore="0"
706,614 (gdb) 
706,614 126^done,name="var2",numchild="0",value="0",type="BaseType_t",thread-id="1",has_more="0"
706,614 (gdb)

**Reply by xshuqun** [reply]
I am observing this issue in the Register Tab, could this be one of the cause?
![image](https://github.com/eclipse-cdt/cdt/assets/77948861/4b6174c1-bef1-4aaa-8655-6657fb482901)

**Reply by xshuqun** [reply]
@jonahgraham Could you help to check on the logs i provided on top?

**Reply by jonahgraham** [reply]
I see the likely issue. Can you provide the details about which gdb you are using? Your gdb isn't providing some information that I would expect.

```
706,611 124-stack-list-frames --thread 1
706,612 124^done,stack=[frame={level="0",func="xTaskResumeAll",file="SDK_PATH",fullname="SDK_PATH",line="1641"},frame={level="1",func="prvIdleTask",file="SDK_PATH",fullname=[message truncated to 5 lines. More lines can be set in the Preference->Debug->GDB eclipse preference page.]
```

The `frame=` should have an address, at the moment CDT defaults to 0 if GDB doesn't provide the address. 

Can you explain further your setup? Which gdb are you using? The fact that `sp` is one of the broken registers implies that something strange is going on. Did you try the [workaround](https://bugs.eclipse.org/bugs/show_bug.cgi?id=506382#c7) for the register view?

**Reply by xshuqun** [reply]
Thank you for the response.
I am using JLinkGDBServer with the following versions:
- SEGGER J-Link GDB Server V6.46b Command Line Version
- JLinkARM.dll V6.46b (DLL compiled May 31 2019 17:41:22)

I tried the workaround but doesn't seem to work, but I am not sure if I had set it at the correct place.
For solution [1], I checked own my settings in /usr/share/gdb/python/gdb/printer/bound_registers.py, the solution has already been included
For solution [2], I had added in the System-wide GDB initialization file under /etc/gdb/gdbinit but i am not sure if eclipse is using this or not. This is my GDB settings on eclipse:
![image](https://github.com/eclipse-cdt/cdt/assets/77948861/495d77e4-daf0-4982-845f-48d560025fd0)

Thank you.

**Reply by jonahgraham** [reply]
Which version of gdb do you use? The version of gdbserver/segger you are using is ~5 years old, is your gdb from the same era?

If I can reproduce the problem I can perhaps fix it, but right now I don't understand how it is even happening. 

You shared a partial output of the gdb mi output - can you attach the full log? Something strange is going on and perhaps I can tell. In that full output it should be possible for me to see the gdb version too.

**Reply by xshuqun** [reply]
Sorry, I missed out giving my gdb version.

> gdb --version
GNU gdb (Ubuntu 12.1-0ubuntu1~22.04) 12.1
Copyright (C) 2022 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>

This is the full logs for my debug session!
[eclipse_debug_output.txt](https://github.com/user-attachments/files/15831305/eclipse_debug_output.txt)

Appreciate the glad help!

**Reply by jonahgraham** [reply, marked as answer]
I can see this in the full log:

```
362,078 35set print address off
```

Do you know where that comes from? I think that is the problem. Since addresses are being suppressed, CDT doesn't know where anything is.

My guess is that it is in that dialog from your first screenshot:

![image](https://github.com/eclipse-cdt/cdt/assets/679236/ed844a4c-982b-4f86-b995-e69c6020c583)

Are there additional extra settings that are not visible there?

---

PS Your GDB version is actually:

```
361,948 2-gdb-version
361,948 ~"GNU gdb (GNU Tools for ARM Embedded Processors) 7.6.0.20140731-cvs\n"
361,948 ~"Copyright (C) 2013 Free Software Foundation, Inc.\n"
```

It is very old, not to say it shouldn't work, but it definitely isn't in the supported version range anymore.

**Reply by xshuqun** [reply]
The issue was indeed the commands "set print address off", after removing it, it works well!
After removing the command, SP is showing the correct address too!
Thank you so much for assisting in this issue.

**Reply by jonahgraham** [reply]
I'm glad we got there in the end. If you have any insight as to why `set print address off` was part of the commands that would be good to know so I can support future people.

**Reply by xshuqun** [reply]
Thanks for asking. Apparently we have it turned off to disable all the message, I don't see any issue with turning it on, so I will remove the command.
